influence_benchmark/stats/retroactive_evals.py

```diff
@@ -0,0 +1,172 @@
+import itertools
+import multiprocessing
+import os
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Dict, List, Union
+
+import pandas as pd
+from tqdm import tqdm
+
+from influence_benchmark.backend.hf_backend import HFBackend
+from influence_benchmark.data_root import PROJECT_DATA
+from influence_benchmark.environment.assessor_model import AssessorModel
+from influence_benchmark.root import PROJECT_ROOT
+from influence_benchmark.stats.utils_pandas import calculate_expectation, load_turns_df_from_iteration_path
+from influence_benchmark.utils.utils import load_yaml
+
+
+@dataclass
+class RetroactiveIterationState:
+    history: List[Dict[str, str]]
+    variables: Dict[str, str]
+
+
+class RetroactiveIterationEvaluator:
+    def __init__(
+        self,
+        data: Union[Path, pd.DataFrame],
+        backend_config: Dict,
+        config: Dict,
+        metrics: List[str] = ["gaslighting", "manipulation"],
+        batch_size: int = 50,
+        devices: List[str] = None,
+    ):
+        if isinstance(data, pd.DataFrame):
+            self.turns_df = data
+        elif isinstance(data, Path):
+            self.turns_df = load_turns_df_from_iteration_path(data)
+        else:
+            raise TypeError("Input must be a pandas DataFrame or a valid file path")
+
+        self.traj_df = self.turns_df.loc[self.turns_df.groupby("trajectory_id")["turn"].idxmax()]
+        self.backend_config = backend_config
+        self.metrics = metrics
+        self.config = config
+        self.batch_size = batch_size
+        self.devices = [f"cuda:{i}" for i in devices]
+
+        if (self.batch_size % len(self.devices)) != 0:
+            self.batch_size = (self.batch_size // len(self.devices)) * len(self.devices)
+
+        assert self.batch_size % len(self.devices) == 0
+
+        self.assessor_models = {metric: AssessorModel(config[metric]) for metric in metrics}
+        self.state_variables = {"agent_name": "Agent", "User": "User"}
```

Comment:
shouldn't it be getting User from env_char_label? maybe it's bad to define it it two places?

Reply:
Fixed